proto/merge.go

```diff
@@ -360,3 +325,52 @@ func mergedFileDescriptors(globalFiles *protoregistry.Files, appFiles map[string
 		File: p.fds,
 	}, nil
 }
+
+// HybridResolver is a protodesc.Resolver that uses both protoregistry.GlobalFiles
+// and the gogo proto global registry, checking protoregistry.GlobalFiles first and
+// then gogo proto global registry.
+var HybridResolver Resolver = &hybridResolver{}
+
+// Resolver is a protodesc.Resolver that can range over all the files in the resolver.
+type Resolver interface {
+	protodesc.Resolver
+
+	// RangeFiles calls f for each file descriptor in the resolver while f returns true.
+	RangeFiles(f func(fileDescriptor protoreflect.FileDescriptor) bool)
+}
+
+type hybridResolver struct{}
+
+var _ protodesc.Resolver = &hybridResolver{}
+
+func (r *hybridResolver) FindFileByPath(path string) (protoreflect.FileDescriptor, error) {
+	if fd, err := protoregistry.GlobalFiles.FindFileByPath(path); err == nil {
+		return fd, nil
+	}
+
+	return gogoProtoRegistry.FindFileByPath(path)
+}
+
+func (r *hybridResolver) FindDescriptorByName(name protoreflect.FullName) (protoreflect.Descriptor, error) {
+	if desc, err := protoregistry.GlobalFiles.FindDescriptorByName(name); err == nil {
+		return desc, nil
+	}
+
+	return gogoProtoRegistry.FindDescriptorByName(name)
+}
+
+func (r *hybridResolver) RangeFiles(f func(fileDescriptor protoreflect.FileDescriptor) bool) {
+	seen := map[protoreflect.FullName]bool{}
```

Comment:
This could be pre-sized to `GlobalFiles.NumFiles()`, but it probably won't be in any hot loops, so it should also be fine as-is.